File: leetcode/1399. 统计最大组的数目.js
/**
 * @param {number} n
 * @return {number}
 */
var countLargestGroup = function(n) {
  const group = new Map()
  const getSum = (num) => {
    let res = 0
    while (num > 1) {
      res += num % 10
      num = Math.floor(num / 10)
    }
    return res + num
    // return [...String(num)].reduce((a, b) => Number(a) + Number(b))
  }
  let max = 1
  for (let i = 1; i <= n; i++) {
    const key = getSum(i)
    if (group.has(key)) {
      const newVal = group.get(key) + 1
      if (newVal > max) {
        max = newVal
      }
      group.set(key, newVal)
    } else {
      group.set(key, 1)
    }
  }
  let res = 0
  for (let val of group.values()) {
    if (val === max) {
      res++
    }
  }
  return res
};

File: leetcode/1399. 统计最大组的数目.cpp
class Solution {
public:
  int countLargestGroup(int n) {
    unordered_map<int, int> m;
    int maxVal = 0;
    for (int i = 1; i <= n; i++) {
      int key = 0;
      int _i = i;
      while (_i) {
        key += _i % 10;
        _i /= 10;
      }
      m[key]++;
      maxVal = max(m[key], maxVal);
    }
    int res = 0;
    for (auto item : m) {
      if (item.second == maxVal) {
        res++;
      }
    }
    return res;
  }
};